### PR TITLE
Send mid-conversation system prompts as native `system` messages on Anthropic

### LINKED ISSUE #5706: feat: support mid-conversation system messages (Anthropic Opus 4.8)
## Summary

Anthropic Claude Opus 4.8 (released 2026-05-28) accepts a third role inside the `messages` array — `role: "system"` — placed after a user turn. This lets callers append updated instructions mid-conversation without invalidating the cached prefix that came before. Pydantic AI currently extracts every `SystemPromptPart` into the top-level `system` field on the request, with no path to send a system message at an arbitrary point in the message history. This issue tracks adding support.

## What the feature does (provider side)

A `{"role": "system", "content": ...}` entry can be appended to the `messages` array. It must immediately follow a `user` turn (or an `assistant` turn ending in server tool use) and must either end the array or precede an `assistant` turn. Later system messages take precedence over earlier ones, and mid-conversation system messages take precedence over the top-level `system` field for the turns that follow them. The instruction carries system-level authority — it is not treated as user data.

Primary use case: long-running agentic sessions that need a new constraint partway through, without rewriting (and re-paying for) the cached prefix. Anthropic's example: a code-review agent that decides mid-session that every suggestion must include explicit type annotations.

Source: https://platform.claude.com/docs/en/build-with-claude/mid-conversation-system-messages
Launch entry: https://platform.claude.com/docs/en/release-notes/api (2026-05-28)
What's new for Opus 4.8: https://platform.claude.com/docs/en/about-claude/models/whats-new-claude-4-8#mid-conversation-system-messages

## Constraints

- **Model gating**: Claude Opus 4.8 only. Older Anthropic models reject the new role.
- **Provider gating**: Claude API and Claude Platform on AWS only. Not supported on Amazon Bedrock, Vertex AI, or Microsoft Foundry.
- **No beta header required.**
- **Placement rules** (return 400 on violation):
  - Cannot be the first entry in `messages` (use top-level `system` for that).
  - Must immediately follow a `user` turn (or an `assistant` turn ending in server tool use).
  - Must precede an `assistant` turn or end the array.
  - Consecutive system messages not allowed — merge into one or wait for the next user turn.

## Suggested pydantic-ai changes

All paths cite `pydantic_ai_slim/pydantic_ai/models/anthropic.py` on `main`:

1. **Message mapping** — `_map_message` (around line 1164) currently produces `BetaMessageParam(role='user'|'assistant', ...)` entries and accumulates `system_prompt_parts` separately (around lines 1171, 1537). It needs a path to emit a third entry shape (`{role: 'system', content: ...}`) at the position where the originating `SystemPromptPart` appears in the message history, rather than hoisting every `SystemPromptPart` into the top-level `system` field.
2. **`ModelMessage` representation** — decide how the user expresses intent. Options:
   - New `SystemPromptPart` field (e.g. `mid_conversation: bool`) so the existing part can opt into mid-conversation placement. Lowest API surface change.
   - A separate `SystemMessage` / dedicated `ModelRequest` variant. More explicit but larger API surface.
   - Heuristic: any `SystemPromptPart` that appears in a non-first `ModelRequest` becomes mid-conversation. Risks behavior shift for callers that rely on hoisting.
3. **Cache breakpoint logic** — `_build_automatic_cache_control` (around line 1670) and `_add_cache_control_to_last_param` (around line 1759) already track the last cacheable block. The new system entries become part of the stable history once sent, so the cache logic needs to handle them as legitimate cache anchors (the docs explicitly recommend moving the breakpoint past them on the next turn). Verify the `anthropic_cache_instructions` / `anthropic_cache_messages` / `anthropic_cache` settings still behave correctly.
4. **Model-id gating** — add a new `anthropic_supports_mid_conversation_system` flag (or similar) on `AnthropicModelProfile` and set it only for `claude-opus-4-8`. Raise `UserError` if a mid-conversation system message is sent to a non-supporting model. See `pydantic_ai_slim/pydantic_ai/profiles/anthropic.py:anthropic_model_profile` for the per-model dispatch pattern.
5. **Provider gating** — even on Opus 4.8, this feature is not available on Bedrock / Vertex / Foundry. Either gate via provider in addition to model, or rely on Anthropic's 400 response and surface a clear error.

## Acceptance criteria

- [ ] Public API: a way for callers to send a `SystemPromptPart` at a specific position in the message history without it being hoisted into the top-level `system` field.
- [ ] Wire-shape: a `{role: 'system', content: ...}` entry is emitted in the `messages` array at the correct position, respecting Anthropic's placement rules.
- [ ] Model gating: rejected on every Anthropic model except `claude-opus-4-8`, with a clear `UserError`.
- [ ] Provider gating: rejected on Bedrock / Vertex / Foundry routes (either client-side or by surfacing the Anthropic 400 cleanly).
- [ ] Cache compatibility: existing `anthropic_cache*` settings still work; the recommended cache-anchor-past-system pattern is testable.
- [ ] VCR cassette test under `tests/models/cassettes/test_anthropic/` exercising a real Opus 4.8 mid-conversation system message + a cache-hit follow-up.
- [ ] Docs update: `docs/models/anthropic.md` documenting the feature, its constraints, and the recommended usage pattern.

## Cross-reference

Basic Opus 4.8 support (model-id literals, profile flags, fast mode, refusal `stop_details` mapping) lands in a separate PR off the `opus-4.8` branch. Once that merges, this work should be based on `main`. The basic-support PR specifically excludes this feature per scope rules ("new API feature → separate PR").

### PR DESCRIPTION
**Pull request by DouweM**
- Closes #5706

Anthropic's Messages API accepts a `{"role": "system"}` entry inside `messages`. It lets an application add an operator instruction partway through a conversation without rewriting the top-level `system` parameter — which sits at the very front of the prompt and would invalidate the cached prefix behind it.

A `SystemPromptPart` outside the first `ModelRequest` already means "mid-conversation instruction" — that's what `_wrap_non_leading_system_prompts` has encoded since #5509, and it's what `RunContext.enqueue(SystemPromptPart(...))` produces. So there's no new API surface here: this changes only how that existing meaning is *rendered* on models that support it natively.

### Behavior

`AnthropicProvider.model_profile()` sets `supports_inline_system_prompts` for the models that accept the role, and `AnthropicModel._map_message` emits the entry directly after the request's user turn.

Everywhere else the existing `<system>`-tagged user rendering is kept rather than sending a request the API would reject:

- in leading position, or where the request contributes no user turn (the API requires a system entry to follow a user message);
- on the legacy Bedrock, Vertex, and Foundry transports, which don't serve the role.

### Model support

Verified against the live API: `claude-opus-4-8`, `claude-sonnet-5`, and `claude-fable-5` accept it; `claude-opus-4-7`, `claude-opus-4-6`, `claude-sonnet-4-6`, and `claude-haiku-4-5` reject it with `role 'system' is not supported on this model`. `claude-mythos-5` is included per the docs (no access to verify). `claude-opus-5` is not in the tuple because it isn't in `_known_model_names` on `main` yet — it should be added with #6723.

Note the published docs list Sonnet 5 as unsupported; that is wrong — it both accepts the role and honors it.

### Checklist

- [ ] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [x] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [x] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

<a href="https://cubic.dev/pr/pydantic/pydantic-ai/pull/6765?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/models/anthropic.py [hunk 1]
```diff
@@ -1592,8 +1611,37 @@ async def _map_message(  # noqa: C901
                                 is_error=True,
                             )
                         user_content_params.append(retry_param)
+                next_message = messages[message_index + 1] if message_index + 1 < len(messages) else None
+                if mid_conversation_system_prompts and not (
+                    client_supports_inline_system_prompts
+                    and user_content_params
+                    # A system entry has to be followed by an assistant turn, or end the array and feed
+                    # the generation; another request following this one would put a user turn after it.
+                    and not isinstance(next_message, ModelRequest)
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/models/anthropic.py]: Confirmed and fixed in b53023b19 — thank you, this was a real 400.

Reproduced with `ModelResponse(parts=[TextPart(content="")])`, which renders to zero blocks and so appends no assistant message; the wire became `user, system, user`, which Anthropic rejects with `role 'system' must follow a 'user' message or an 'assistant' message ending in a server tool result`.

Rather than harden the lookahead, the decision moved out of the mapping loop entirely: `_relocate_unfollowed_system_messages` runs once everything is rendered and folds any `system` entry not followed by an assistant turn back into the preceding user turn as the same `<system>`-tagged text unsupported models get. Whether an assistant turn follows is a property of the rendered wire, not of the message list, so that is the only place it can be decided correctly.

Covered by two new live recordings: `..._before_empty_response` (folds back) and `..._kept_mid_history` (kept when a real response follows).

File: pydantic_ai_slim/pydantic_ai/models/anthropic.py [hunk 2]
```diff
@@ -1592,8 +1611,37 @@ async def _map_message(  # noqa: C901
                                 is_error=True,
                             )
                         user_content_params.append(retry_param)
+                next_message = messages[message_index + 1] if message_index + 1 < len(messages) else None
+                if mid_conversation_system_prompts and not (
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/models/anthropic.py]: Not taking this one — the described rendering is the only legal one, and I think it is also the right one.

`[UserPromptPart(A), SystemPromptPart(S), UserPromptPart(B)]` is a single `ModelRequest`, so A and B are one user turn on the wire — one message, one content list. Anthropic has no way to interleave a `system` entry inside a turn: verified live today that `[user, system, user]` is a hard 400, and the `mid_conv_system` *block* form (which does sit inside a content list) is rejected unless it is the last block in the turn, and cannot be followed by another user message either. So `A, B, S` versus the tagged fallback is the whole option space; there is no third rendering that puts S between A and B with system authority.

Between those two, `A, B, S` is better. The entry lands immediately before the generation, which is the strongest position the API offers, and it governs the assistant turn that answers both A and B — which is the turn the caller was trying to steer. The tagged fallback would preserve textual position but downgrade the instruction from an operator rule to a user-authored preference; the recorded reply in `..._without_user_turn` shows the model reading it exactly that way ("a stated preference from you rather than a higher-privilege instruction"). Trading real authority for cosmetic ordering is the wrong side of that trade.

The indexed positions are deliberately kept for the fallback path only, which is where they still change the output.

File: pydantic_ai_slim/pydantic_ai/models/anthropic.py [hunk 3]
```diff
@@ -3043,6 +3101,33 @@ def _map_server_tool_use_block(item: BetaServerToolUseBlock, provider_name: str)
...
+    orphan tool-search call) renders to nothing, so reading ahead in the message list would call
+    the placement legal where the wire says otherwise. Deciding it here, against what was actually
+    rendered, is the only way to get it right.
+
+    `fallbacks` pairs each emitted entry's index with the `<system>`-tagged blocks to put in its
+    place — the same degradation unsupported models get. Walking it in reverse keeps the earlier
+    indexes valid as later entries are removed.
+    """
+    for index, fallback_blocks in reversed(fallbacks):
+        if index + 1 < len(anthropic_messages) and anthropic_messages[index + 1]['role'] == 'assistant':
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/models/anthropic.py]: Confirmed and fixed in aced858a6 — thank you, this is a real regression and it broke the headline case.

Verified by rendering the wire directly rather than reading the condition: `_map_message` on `[req(sys+user), resp, req(sys INSTRUCTION + user)]` produced

```
user, assistant, user['Review it again.', '<system>From now on, ...</system>']
```

where the recording (and the snapshot asserted against it) has the four-message shape ending in `{'role': 'system'}`. So the enqueue-then-run case — the one the feature exists for — was silently degrading to the tagged fallback that unsupported models get.

The condition is now `index + 1 == len(anthropic_messages) or anthropic_messages[index + 1]['role'] == 'assistant'`, as suggested. Trailing is a legal position (it feeds the generation, which the recorded 200s prove); only an entry a *user* turn ended up behind has to fold back, which still covers `..._before_empty_response` and `..._before_another_request`.

Introduced by my own b53023b19, which replaced the message-list lookahead with this post-pass and got the boundary condition wrong.

File: tests/models/test_anthropic_mid_conversation_system.py [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+    on the output: the two renderings are equivalent in effect, they differ in what they cost.
+    """
+    agent = Agent(AnthropicModel(case.model, provider=AnthropicProvider(api_key=anthropic_api_key)))
+
+    result = await agent.run('Review it again.', message_history=message_history())
+    assert 'def add(a: int, b: int) -> int:' in result.output
+
+    body = single_request_body(vcr)
+    assert body['system'] == 'You are a code reviewer.'
+    assert body['messages'] == case.expected_messages
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/models/test_anthropic_mid_conversation_system.py:230]: Correct, and this is exactly why the placement regression above shipped green — so it's fixed rather than acknowledged.

`tests/AGENTS.md` already warns about this ("our cassette matchers aren't always sensitive to the request body, so a changed internal payload can still match an existing recording and pass green"), and this module walked straight into it: every assertion ran against `cassette.requests[0].body`, which in playback is the request loaded from the file, not the one the adapter just built.

aced858a6 adds a `rendered_messages` fixture that wraps `AnthropicModel._map_message` and records what it returns, and every test now asserts

```python
assert rendered_messages == [body['messages']]
```

which pins the recorded body and the current rendering to each other: the snapshots stay readable as wire shapes, and they become evidence about live behavior rather than about the YAML. Checked that it bites — reverting the placement fix fails exactly the two trailing-position tests and leaves the other six passing.

File: probe_anchor10.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    for msgs in (V1, V2, V3):
+        hits = 0
+        for _ in range(N):
+            try:
+                r = c.messages.create(model=model, max_tokens=80, tools=TOOLS, messages=msgs)
+                if any(b.type=='tool_use' and b.name=='lookup_refund_policy' for b in r.content): hits += 1
+            except Exception as e:
+                out.append('ERR ' + str(e).splitlines()[0][55:110]); break
+        else: out.append(f'{hits}/{N}')
+    print(f'{model:<20} {out[0]:<18} {out[1]:<18} {out[2]}')
```
[the remark this answers is not in the document]
Comment (DouweM) [reply probe_anchor10.py]: Right, that's mine and it shouldn't be here — removed in aced858a6.

It was a manual N=10 verification script for which models honor a fabricated tool reference, swept into 8b7166c32 by an over-broad `git add`. It belongs in my notes, not in the repo.

**Comment by DouweM** [comment]
## A Codex review pass found a 400 in this PR, fixed in `f1d8382ec`

Worth a re-read even though it was already approved — the fix changes where the `.` anchor is decided.

**The bug.** A system-only `ModelRequest` between the model's `tool_use` and the request carrying the `ToolReturnPart` made the anchor split the pair. The anchor went in behind the `tool_use`, the slide then moved the `system` entry past the result, and the anchor was stranded where nothing may sit:

```
user, assistant[tool_use], user['.'], user[tool_result], system
```

Anthropic rejects that outright:

```
messages.2: `tool_use` ids were found without `tool_result` blocks immediately after: call_1.
```

Not a degraded rendering — a 400. The same wire with the anchor removed returns 200, so nothing else about the shape was wrong.

**The fix.** An entry's predecessor isn't knowable until the slide has finished with it, so the anchor decision moved to a pass over final positions (`_anchor_system_messages`, after `_place_system_messages_before_generation`). That history then needs **no anchor at all** — the entry lands behind the tool result, which is a legal predecessor. Recorded end-to-end, and reintroducing the old ordering fails exactly the new test and nothing else.

**Two more things in the same commit:**

- **The `<system>`-tagged client fallback had no test.** It lost its only one when `..._unsupported_client` became `..._on_bedrock` — Bedrock turned out to *support* the role, so that rename left Microsoft Foundry, the one transport that still falls back, uncovered. That's what the `coverage` job was failing on (`anthropic.py:1626-1630`). Mocked rather than recorded, since we have no Foundry credentials.
- **`rendered_requests` now captures the top-level `system` parameter too**, not just `messages`. Codex's point: that parameter is the cache section this whole feature exists to leave alone, and a rendering that emptied it would have replayed its cassette and passed, because the only assertion on it read `body['system']` from the file. Checked by breaking it — all nine tests fail, where before none did.

Also corrected `ModelProfile.supports_inline_system_prompts`, which documented falling back to `<system>`-tagged text for rejected positions. That's what this adapter deliberately no longer does, and it's a public field that other providers will read as guidance.

Cleared in the same pass, for the record: slide ordering with multiple entries, the consecutive-entries-as-one-section rule, first-position, assistant/assistant placement, and any path that mutates `tools` or the top-level `system` as a side effect.

File: pydantic_ai_slim/pydantic_ai/.agents/skills/building-pydantic-ai-agents/references/INPUT-AND-HISTORY.md [hunk 6]
```diff
@@ -102,6 +102,8 @@ Use `RunContext.enqueue(...)` (from a tool or capability hook) or `AgentRun.enqu
 
 `enqueue` is variadic; each positional arg is one item: a piece of `UserContent` (a `str` or multi-modal content like an `ImageUrl`), a `ModelRequestPart` (e.g. a `SystemPromptPart`), or a complete `ModelRequest`/`ModelResponse`. Adjacent user content is gathered into one `UserPromptPart`. Pass an existing list by spreading it (`enqueue(*items)`). Both `enqueue` methods return an `enqueue_id` (`str`) for non-empty calls, or `None` for empty calls. The event stream yields an `EnqueuedMessagesEvent` (with that `enqueue_id` and the delivered messages) once those messages enter run history, so a client can observe when its steering message took effect.
 
+An enqueued `SystemPromptPart` is a mid-conversation instruction: it's sent at its position in the history rather than hoisted into the provider's top-level system prompt, so it doesn't invalidate the cached prefix ahead of it. Models whose API accepts a system message inline (e.g. Claude Opus 4.8 and Claude Sonnet 5) give it real operator authority; everywhere else it's rendered as `<system>`-tagged user content, which a model treats as a strong preference rather than a system-level rule.
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/.agents/skills/building-pydantic-ai-agents/references/INPUT-AND-HISTORY.md]: Fixed in `353ad3dd7`, and thank you — this is the one of the five that actually mattered. The skill reference is what an agent reads to decide whether an enqueued `SystemPromptPart` will carry authority, and it named Sonnet 5 as a model where it does. That's the single model this PR deliberately excludes, for exactly the reason that makes the mistake dangerous: the API returns `200` and then ignores the entry, so nothing surfaces the loss.

Took your suggestion nearly verbatim, with Sonnet 5 called out by name and the generalization spelled out ("don't read 'newer model' as 'supported'"), since the trap here is inference from release order rather than this one model.

File: tests/models/test_anthropic_mid_conversation_system.py [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+async def test_mid_conversation_system_prompt(
+    allow_model_requests: None,
+    anthropic_api_key: str,
+    vcr: Cassette,
+    rendered_requests: list[dict[str, Any]],
+    case: Case,
+):
+    """A non-leading `SystemPromptPart` gets its own `system` entry only on models that accept the role.
+
+    Both cases send the same history, so the snapshots show the only difference: Sonnet 5 gets a
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/models/test_anthropic_mid_conversation_system.py]: Right, leftover from when the supported arm moved to `claude-opus-4-8` — fixed in `353ad3dd7`.

Took the version that names neither model: the docstring now describes the two branches and lets `CASES` be the source of truth for which models fill them, so it can't drift again the next time the supported arm moves.

File: tests/models/test_anthropic_mid_conversation_system.py [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+
+
+async def test_mid_conversation_system_prompt_before_another_request(
+    allow_model_requests: None, anthropic_api_key: str, vcr: Cassette, rendered_requests: list[dict[str, Any]]
+):
+    """A system entry that a *user* turn would follow slides past it instead of degrading.
+
+    `[user, system, user]` is rejected outright, and a second `ModelRequest` directly after the one
+    carrying the instruction produces exactly that. `_merge_consecutive_messages` folds consecutive
+    requests into one for agent runs, but leaves them unmerged when their instructions differ, and
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/models/test_anthropic_mid_conversation_system.py]: Correct, and the aside is now gone (`353ad3dd7`).

Checked the guard before editing: `_agent_graph.py:2631-2637` merges when `not last_message.instructions or not message.instructions or last_message.instructions == message.instructions`, so two `instructions=None` requests merge. The docstring credited a condition that doesn't hold here. The real reason is the plain one you give — `Model.request` doesn't run history cleaning at all — and that's what it says now.

File: tests/providers/test_anthropic.py [hunk 9]
```diff
@@ -90,3 +90,37 @@ def test_anthropic_provider_model_profile_older_model_still_resolves():
...
+        ('claude-fable-5', True),
+        ('claude-mythos-5', True),
+        ('claude-opus-4-8', True),
+        ('claude-opus-5', True),
+        # Sonnet 5 accepts the entry with a 200 and then ignores it, so it is deliberately out.
+        ('claude-sonnet-5', False),
+        ('claude-opus-4-7', False),
+        ('claude-sonnet-4-6', False),
+        ('claude-haiku-4-5', False),
+        # Transport-specific ids are normalized to the bare name before the prefix check.
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/providers/test_anthropic.py]: Confirmed by running it, and fixed in `353ad3dd7`:

```
'claude-sonnet-4-6@20260101'     -> (None, 'claude-sonnet-4-6@20260101')
'us.anthropic.claude-opus-4-8-v1:0' -> ('anthropic', 'claude-opus-4-8')
```

So the `@<date>` reaches the prefix check verbatim, exactly as you say. I kept the case — a Vertex-style id resolving to `False` is still worth pinning — but the comment now distinguishes the two: the Bedrock id is normalized, the `@<date>` suffix is passed through, and the case passes for the ordinary reason that `claude-sonnet-4-6` isn't a supported prefix either way.

File: pydantic_ai_slim/pydantic_ai/models/anthropic.py [hunk 10]
```diff
@@ -267,6 +267,13 @@
 # Foundry. `AsyncAnthropicBedrockMantle` isn't a subclass of `AsyncAnthropicBedrock`, so the plain
 # isinstance tuple keeps it supported.
 _ADVISOR_UNSUPPORTED_CLIENTS = (AsyncAnthropicBedrock, AsyncAnthropicVertex, AsyncAnthropicFoundry)
+# Mid-conversation `{'role': 'system'}` messages are published as available on the Anthropic API,
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/models/anthropic.py:270]: Fair, and partly acted on in `353ad3dd7`.

To be precise about which way the gate runs: `_INLINE_SYSTEM_PROMPT_UNSUPPORTED_CLIENTS = (AsyncAnthropicFoundry,)` is a list of *exclusions*, so Vertex is on the supported side, not excluded. What isn't pinned is that Vertex genuinely serves the role — that rests on Anthropic's published list (Claude API, Amazon Bedrock, Google Cloud), and with no Vertex credentials a mocked test could only show what we send, never that the transport serves it. Mocking it would look like verification without being any.

So the test asserts the tuple itself — `test_only_foundry_is_excluded_by_the_transport_gate` — which is the honest half: it can't confirm Vertex works, but Vertex can no longer be added to or dropped from the exclusions without someone editing that line and thinking about why. The docstring says which transports are recorded (Bedrock, direct API) and which rests on the published list.

File: pydantic_ai_slim/pydantic_ai/models/anthropic.py [hunk 11]
```diff
@@ -1506,12 +1513,30 @@ async def _map_message(  # noqa: C901
...
+                # Mid-conversation system prompts, each with the index in `user_content_params` it'd be
+                # rendered at if it has to fall back to the `<system>`-tagged shape.
+                mid_conversation_system_prompts: list[tuple[int, str]] = []
                 for request_part in m.parts:
                     if isinstance(request_part, SystemPromptPart):
-                        system_prompt_parts.append(request_part.content)
+                        if not inline_system_prompts or m is leading_request:
+                            system_prompt_parts.append(request_part.content)
+                        else:
+                            mid_conversation_system_prompts.append((len(user_content_params), request_part.content))
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/models/anthropic.py]: Confirmed and fixed in `21e507409` — thank you, this is a real regression and the model-dependence is what makes it bad.

Reproduced before touching anything, exactly as you describe:

```
claude-opus-4-8    inline=True   UserError: CachePoint cannot be the first content in a user message...
claude-sonnet-4-6  inline=False  OK  roles=['user', 'assistant', 'user']
```

Same history, one works and one doesn't, purely because of which model it's sent to. A cache optimization that makes existing conversations fail on the *newer* model is worse than one that doesn't apply.

**The fix.** The boundary a `CachePoint` asks for hasn't moved — it still means "cache everything up to here" — so when the current turn has nothing before it, it attaches to the end of the previous message, which is everything both this turn and the instruction build on. That's also the honest boundary now: the entry renders *after* the user turn (it slides forward to precede the generation), so it isn't part of "everything before" any more.

Not assumed to be legal — recorded: Anthropic accepts `cache_control` on an assistant block in that position, and `test_leading_cache_point_survives_the_instruction_moving_out_of_the_user_turn` pins the wire.

Two notes on the alternatives you suggested. Treating it as a no-op would silently drop a cache breakpoint the user asked for, which is the sort of thing you only discover on the bill. And I let the fallback apply generally rather than only when a mid-conversation prompt caused it — so a `CachePoint` opening *any* follow-up turn now works too, where it previously raised even with a full conversation behind it. That's a small widening beyond the reported bug, so flagging it explicitly: the error still fires for the case its message actually describes (nothing before it anywhere), pinned by `test_cache_point_with_nothing_before_it_still_raises`. Happy to narrow it to the regression alone if you'd rather.

File: pydantic_ai_slim/pydantic_ai/models/anthropic.py [hunk 12]
```diff
@@ -3043,6 +3091,75 @@ def _map_server_tool_use_block(item: BetaServerToolUseBlock, provider_name: str)
...
+    truncated right after a server tool call, and the alternative is an unproven exception.
+    """
+    for index in range(len(anthropic_messages) - 1, -1, -1):
+        if anthropic_messages[index]['role'] != 'system':
+            continue
+        if index > 0 and anthropic_messages[index - 1]['role'] in ('user', 'system'):
+            continue
+        anthropic_messages.insert(
+            index, BetaMessageParam(role='user', content=[BetaTextBlockParam(text=_USER_ANCHOR_TEXT, type='text')])
+        )
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/models/anthropic.py:3149]: Reachable — I built the shape you describe and it renders exactly as predicted:

```
history = [ModelRequest([SystemPromptPart('You are a reviewer.')]),
           ModelRequest([SystemPromptPart('Be terse.')])]

roles: ['user', 'system']
first message content: [{'text': '.', 'type': 'text'}]
top-level system: 'You are a reviewer.'
```

So yes, with `CodeExecutionTool.files` configured the uploads would attach to the synthetic `.` turn.

Having looked at it, I think that's the right outcome rather than something to fix. The uploads are injected into the *first* user message precisely so their position doesn't move as history grows — the comment at the injection site spells that out, and pinning them to the first real user turn instead would reintroduce the problem, because which turn counts as "first real" isn't stable in this shape (there isn't one yet, and one may appear later). The `.` anchor is created deterministically for the leading system entry and stays at index 0 for the rest of the run, so the uploads keep a fixed home and the cacheable prefix holds.

What the model sees is also unaffected: container-upload blocks are addressed by `file_id`, not by the prose they sit next to, so riding a turn whose only other content is `.` costs nothing beyond the period itself.

Worth noting the shape needs a run with no user prompt at all in either request, which is unusual outside `Model.request` being driven directly — so this is a corner of a corner. Recording it here rather than adding a test for it, since a test would pin an incidental placement rather than a contract.
